File: 01_Linked_Lists/07_Infix_to_Postfix/main.cpp
#include<iostream>
#include<string>
#include<stack>

// Utility Function
bool isOperand(char ch){
    return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
}

int prec(char ch){  // Function to return the precedence of the operator
    switch(ch){
        case '+':
        case '-':
            return 1;
        case '*':
        case '/':
            return 2;
        case '^':
            return 3;
    }
    return -1;
}

// Main Function
void infix_to_postifx(std::string s){
    std::stack<char> stack;
    stack.push('$');
    int len = s.length();
    std::string ns;

    for(int i = 0; i < len; i++){
        // If the scanned character is an operand
        if(isOperand(s[i]))
            ns += s[i];
        // If the scanned character is '('
        else if(s[i] == '(')
            stack.push(s[i]);
        // If the scanned operator is ')'
        else if(s[i] == ')'){
            while(stack.top() != '$' && stack.top() != '('){
                char c = stack.top();
                stack.pop();
                ns += c;
            }
            if(stack.top() == '(')
                stack.pop();
        }
        // If the operator is scanned
        else{
            while(stack.top() != '$' && prec(s[i]) <= prec(stack.top())){
                char c = stack.top();
                stack.pop();
                ns += c;
            }           
            stack.push(s[i]);     
        }
    }

    while(stack.top() != '$'){
        char c = stack.top();
        stack.pop();
        ns += c;
    }

    std::cout<<ns<<std::endl;
}

// Driver Program
int main(){
    std::string exp = "a+b*(c^d-e)^(f+g*h)-i";
    infix_to_postifx(exp);
    return 0;
}

File: 01_Linked_Lists/07_Infix_to_Postfix/main.py
# Class to convert the expression
class Conversion:
    # Constructor to initialise the class object
    def __init__(self, cap):
        self.top = -1
        self.capacity = cap

        # This list is used as a stack
        self.array = []
       
        # List for the output string
        self.output = []

        # Precedence
        self.precedence = {'+':1, '-':1, '*':2, '/':2, '^':3}

    # Utiltiy function
    def isEmpty(self):  # To check if the stack is empty
        return True if self.top == -1 else False
    
    def isOperand(self, ch):    # to check the scanned character is an operand
        return ch.isalpha()

    def notGreater(self, ch):   # Check if the precendence of operator is less
        try:
            a = self.precedence[ch]
            b = self.precedence[self.peek()]
            return True if a <=b else False
        except KeyError:
            return False


    # Stack methods
    def peek(self):   # Return the value of the top of the stack
        return self.array[-1]
    
    def pop(self):  # Pop the element from the stack
        if not self.isEmpty():
            self.top -= 1
            return self.array.pop()
        else:
            return "$" 
        
    def push(self, op): # Push the element to the stack
        self.top += 1
        self.array.append(op)


    # Main function
    def infix_to_postfix(self, exp):
        # Iterate over the exp
        for ch in exp:
            # If the character is operand, add it to output
            if self.isOperand(ch):
                self.output.append(ch)
            
            # If the character is an '(', push it to the stack
            elif ch == '(':
                self.push(ch)
            
            # If the character is an ')
            elif ch == ')':
                while((not self.isEmpty()) and self.peek() != '('):
                    self.output.append(self.pop())
                if(not self.isEmpty() and self.peek() != '('):
                    return
                else:
                    self.pop()

            # An operator is scanned
            else:
                while(not self.isEmpty() and self.notGreater(ch)):
                    self.output.append(self.pop())
                self.push(ch)
            
        # Pop all the operator from the stack
        while not self.isEmpty():
            self.output.append(self.pop()) 

        print("".join(self.output))


# Driver Program
exp = "a+b*(c^d-e)^(f+g*h)-i"
obj = Conversion(len(exp))
obj.infix_to_postfix(exp)